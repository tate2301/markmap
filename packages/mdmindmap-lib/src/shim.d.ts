declare interface Window {
  WebFontConfig?: unknown;
  katex?: any;
  Prism?: any;
  hljs?: typeof import('highlight.js').default;
}

declare module '*.svg?raw' {
  export default string;
}
