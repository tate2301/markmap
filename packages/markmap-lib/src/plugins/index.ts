import pluginCheckbox from './checkbox';
import pluginFrontmatter from './frontmatter';
import pluginHljs from './hljs';
import pluginKatex from './katex';
import pluginNpmUrl from './npm-url';
import pluginSourceLines from './source-lines';
import pluginBidirectional from './bidirectional';

export * from './base';

export {
  pluginCheckbox,
  pluginFrontmatter,
  pluginHljs,
  pluginKatex,
  pluginNpmUrl,
  pluginSourceLines,
  pluginBidirectional,
};

export const plugins = [
  pluginFrontmatter,
  pluginKatex,
  pluginHljs,
  pluginNpmUrl,
  pluginCheckbox,
  pluginSourceLines,
  pluginBidirectional,
];
