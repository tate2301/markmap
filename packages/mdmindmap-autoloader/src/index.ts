import {
  buildCSSItem,
  buildJSItem,
  loadCSS,
  loadJS,
  urlBuilder,
} from 'mdmindmap-common';
import type { Transformer } from 'mdmindmap-lib';
import type { AutoLoaderOptions } from './types';

export * from './types';

const enabled: Record<string, boolean> = {};

const autoLoaderOptions = {
  baseJs: [
    `d3@${__define__.D3_VERSION}`,
    `mdmindmap-lib@${__define__.LIB_VERSION}`,
    `mdmindmap-view@${__define__.VIEW_VERSION}`,
    `mdmindmap-toolbar@${__define__.TOOLBAR_VERSION}`,
  ],
  baseCss: [`mdmindmap-toolbar@${__define__.TOOLBAR_VERSION}/dist/style.css`],
  manual: false,
  toolbar: false,
  ...(window.markmap?.autoLoader as Partial<AutoLoaderOptions>),
};

async function initialize() {
  if (typeof autoLoaderOptions.provider === 'function') {
    urlBuilder.setProvider(
      (urlBuilder.provider = 'autoLoader'),
      autoLoaderOptions.provider,
    );
  } else if (typeof autoLoaderOptions.provider === 'string') {
    urlBuilder.provider = autoLoaderOptions.provider;
  } else {
    try {
      await urlBuilder.findFastestProvider();
    } catch {
      // ignore
    }
  }
  await Promise.all([
    loadJS(
      autoLoaderOptions.baseJs.map((item) =>
        typeof item === 'string'
          ? buildJSItem(urlBuilder.getFullUrl(item))
          : item,
      ),
    ),
    loadCSS(
      autoLoaderOptions.baseCss.map((item) =>
        typeof item === 'string'
          ? buildCSSItem(urlBuilder.getFullUrl(item))
          : item,
      ),
    ),
  ]);
  const { markmap } = window;
  const style = document.createElement('style');
  style.textContent = markmap.globalCSS;
  // Insert global CSS to body so it has higher priority than prism.css, etc.
  document.body.prepend(style);
  autoLoaderOptions.onReady?.();
}

export const ready = initialize();

function transform(transformer: Transformer, content: string) {
  const result = transformer.transform(content);
  const keys = Object.keys(result.features).filter((key) => !enabled[key]);
  keys.forEach((key) => {
    enabled[key] = true;
  });
  const { styles, scripts } = transformer.getAssets(keys);
  const { markmap } = window;
  if (styles) markmap.loadCSS(styles);
  if (scripts) markmap.loadJS(scripts);
  return result;
}

export function render(el: HTMLElement) {
  const { Transformer, Markmap, deriveOptions, Toolbar } = window.markmap;
  const lines = el.textContent?.split('\n') || [];
  let indent = Infinity;
  lines.forEach((line) => {
    const spaces = line.match(/^\s*/)?.[0].length || 0;
    if (spaces < line.length) indent = Math.min(indent, spaces);
  });
  const content = lines
    .map((line) => line.slice(indent))
    .join('\n')
    .trim();
  const transformer = new Transformer(autoLoaderOptions.transformPlugins);
  transformer.urlBuilder = urlBuilder;
  el.innerHTML = '<svg></svg>';
  const svg = el.firstChild as SVGElement;
  const mm = Markmap.create(svg, { embedGlobalCSS: false });
  if (autoLoaderOptions.toolbar) {
    const { el: toolbar } = Toolbar.create(mm);
    Object.assign(toolbar.style, {
      position: 'absolute',
      right: '20px',
      bottom: '20px',
    });
    el.append(toolbar);
  }
  const doRender = async () => {
    const { root, frontmatter } = transform(transformer, content);
    const markmapOptions = frontmatter?.markmap;
    const frontmatterOptions = deriveOptions(markmapOptions);
    await mm.setData(root, frontmatterOptions);
    mm.fit();
  };
  transformer.hooks.retransform.tap(doRender);
  doRender();
}

export async function renderAllUnder(container: ParentNode) {
  await ready;
  container.querySelectorAll<HTMLElement>('.markmap').forEach(render);
}

export function renderAll() {
  return renderAllUnder(document);
}

if (!autoLoaderOptions.manual) {
  if (document.readyState === 'loading')
    document.addEventListener('DOMContentLoaded', () => {
      renderAll();
    });
  else renderAll();
}
