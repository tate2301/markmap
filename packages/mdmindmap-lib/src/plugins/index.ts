import pluginCheckbox from './checkbox';
import pluginFrontmatter from './frontmatter';
import pluginHljs from './hljs';
import pluginKatex from './katex';
import pluginNpmUrl from './npm-url';
import pluginSourceLines from './source-lines';

export * from './base';

export {
  pluginCheckbox,
  pluginFrontmatter,
  pluginHljs,
  pluginKatex,
  pluginNpmUrl,
  pluginSourceLines,
};

export const plugins = [
  pluginFrontmatter,
  pluginKatex,
  pluginHljs,
  pluginNpmUrl,
  pluginCheckbox,
  pluginSourceLines,
];
