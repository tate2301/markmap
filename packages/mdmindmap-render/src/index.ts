import {
  IAssets,
  IPureNode,
  JSItem,
  UrlBuilder,
  buildJSItem,
  urlBuilder as defaultUrlBuilder,
  persistCSS,
  persistJS,
} from 'mdmindmap-common';
import type { IMarkmapJSONOptions, IMarkmapOptions } from 'mdmindmap-view';

export const template = __define__.TEMPLATE || '';

export const baseJsPaths = [
  `d3@${__define__.D3_VERSION}/dist/d3.min.js`,
  `mdmindmap-view@${__define__.VIEW_VERSION}/dist/browser/index.js`,
];

export function fillTemplate(
  root: IPureNode | null,
  assets: IAssets,
  extra?: {
    baseJs?: JSItem[];
    jsonOptions?: Partial<IMarkmapJSONOptions>;
    getOptions?: (
      jsonOptions: Partial<IMarkmapJSONOptions>,
    ) => Partial<IMarkmapOptions>;
    /** See https://github.com/gera2ld/npm2url */
    urlBuilder?: UrlBuilder;
  },
): string {
  extra = {
    ...extra,
  };
  const urlBuilder = extra.urlBuilder ?? defaultUrlBuilder;
  extra.baseJs ??= baseJsPaths
    .map((path) => urlBuilder.getFullUrl(path))
    .map((path) => buildJSItem(path));
  const { scripts, styles } = assets;
  const cssList = [...(styles ? persistCSS(styles) : [])];
  const context = {
    getMarkmap: () => window.markmap,
    getOptions: extra.getOptions,
    jsonOptions: extra.jsonOptions,
    root,
  };
  const jsList = [
    ...persistJS(
      [
        ...extra.baseJs,
        ...(scripts || []),
        {
          type: 'iife',
          data: {
            fn: (
              getMarkmap: (typeof context)['getMarkmap'],
              getOptions: (typeof context)['getOptions'],
              root: (typeof context)['root'],
              jsonOptions: IMarkmapJSONOptions,
            ) => {
              const markmap = getMarkmap();
              window.mm = markmap.Markmap.create(
                'svg#mindmap',
                (getOptions || markmap.deriveOptions)(jsonOptions),
                root,
              );
            },
            getParams: ({ getMarkmap, getOptions, root, jsonOptions }) => {
              return [getMarkmap, getOptions, root, jsonOptions];
            },
          },
        } as JSItem,
      ],
      context,
    ),
  ];
  const html = template
    .replace('<!--CSS-->', () => cssList.join(''))
    .replace('<!--JS-->', () => jsList.join(''));
  return html;
}
