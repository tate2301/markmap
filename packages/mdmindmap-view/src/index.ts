export { loadJS, loadCSS } from 'mdmindmap-common';

export * from './types';
export * from './constants';
