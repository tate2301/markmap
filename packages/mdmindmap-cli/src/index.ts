import { readFile, writeFile } from 'fs/promises';
import { CSSItem, JSItem, buildJSItem, mergeAssets } from 'mdmindmap-common';
import {
  Transformer,
  type IAssets,
  type IMarkmapCreateOptions,
} from 'mdmindmap-lib';
import { baseJsPaths, fillTemplate } from 'mdmindmap-render';
import open from 'open';
import { resolve } from 'path';
import { IDevelopOptions } from './types';
import {
  ASSETS_PREFIX,
  config,
  localProvider,
  toolbarAssets,
} from './util/common';

export * from './types';
export * from './util/dev-server';
export { config };

export * as markmap from 'mdmindmap-lib';

async function loadFile(path: string) {
  if (path.startsWith(ASSETS_PREFIX)) {
    const relpath = path.slice(ASSETS_PREFIX.length);
    return readFile(resolve(config.assetsDir, relpath), 'utf8');
  }
  const res = await fetch(path);
  if (!res.ok) throw res;
  return res.text();
}

async function inlineAssets(assets: IAssets): Promise<IAssets> {
  const [scripts, styles] = await Promise.all([
    Promise.all(
      (assets.scripts || []).map(
        async (item): Promise<JSItem> =>
          item.type === 'script' && item.data.src
            ? {
                type: 'script',
                data: {
                  textContent: await loadFile(item.data.src),
                },
              }
            : item,
      ),
    ),
    Promise.all(
      (assets.styles || []).map(
        async (item): Promise<CSSItem> =>
          item.type === 'stylesheet'
            ? {
                type: 'style',
                data: await loadFile(item.data.href),
              }
            : item,
      ),
    ),
  ]);
  return {
    scripts,
    styles,
  };
}

export async function createMarkmap(
  options: IMarkmapCreateOptions & IDevelopOptions & { open: boolean },
): Promise<void> {
  const transformer = new Transformer();
  if (options.offline) {
    transformer.urlBuilder.setProvider('local', localProvider);
    transformer.urlBuilder.provider = 'local';
  } else {
    try {
      await transformer.urlBuilder.findFastestProvider();
    } catch {
      // ignore
    }
  }
  const { root, features, frontmatter } = transformer.transform(
    options.content || '',
  );
  const otherAssets = mergeAssets(
    {
      scripts: baseJsPaths.map(buildJSItem),
    },
    options.toolbar ? toolbarAssets : null,
  );
  let assets = mergeAssets(
    {
      scripts: otherAssets.scripts?.map((item) => transformer.resolveJS(item)),
      styles: otherAssets.styles?.map((item) => transformer.resolveCSS(item)),
    },
    transformer.getUsedAssets(features),
  );
  if (options.offline) assets = await inlineAssets(assets);
  const html = fillTemplate(root, assets, {
    baseJs: [],
    jsonOptions: (frontmatter as any)?.markmap,
    urlBuilder: transformer.urlBuilder,
  });
  const output = options.output || 'markmap.html';
  await writeFile(output, html, 'utf8');
  if (options.open) open(output);
}
