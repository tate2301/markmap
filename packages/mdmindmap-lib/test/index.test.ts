import { wrapFunction } from 'mdmindmap-common';
import { expect, test } from 'vitest';
import { Transformer, builtInPlugins } from '../src/index';

test('plugins', () => {
  const transformer = new Transformer();
  expect(transformer.plugins.map((plugin) => plugin.name)).toEqual([
    'frontmatter',
    'katex',
    'hljs',
    'npmUrl',
    'checkbox',
    'sourceLines',
  ]);
  const assets = transformer.getAssets();
  expect(assets).toMatchSnapshot();
});

test('custom url provider', () => {
  const transformer = new Transformer();
  transformer.urlBuilder.setProvider('local', (path) => `/local/${path}`);
  transformer.urlBuilder.provider = 'local';
  const assets = transformer.getAssets();
  expect(assets).toMatchSnapshot();
});

test('content without frontmatter', () => {
  const transformer = new Transformer();
  const result = transformer.transform(`\
- l1
  - l1.1
  - l1.2
    - l1.2.1
`);
  expect(result).toMatchSnapshot();
});

test('content with frontmatter', () => {
  const transformer = new Transformer();
  const result = transformer.transform(`\
---
markmap:
  color: blue
---

- l1
  - l1.1
  - l1.2
    - l1.2.1
`);
  expect(result).toMatchSnapshot();
});

test('content with line endings of CRLF', () => {
  const transformer = new Transformer();
  const result = transformer.transform(
    `\
---
markmap:
  color: blue
---

- l1
  - l1.1
  - l1.2
    - l1.2.1
`.replace(/\n/g, '\r\n'),
  );
  expect(result).toMatchSnapshot();
});

test('content with only katex enabled', () => {
  const transformer = new Transformer();
  const result = transformer.transform(`\
---
markmap:
  color: blue
---

- $x = {-b \\pm \\sqrt{b^2-4ac} \\over 2a}$
`);
  expect(result).toMatchSnapshot();
  expect(transformer.getUsedAssets(result.features)).toMatchSnapshot();
});

test('tables', () => {
  const transformer = new Transformer();
  const result = transformer.transform(`\
| products | price |
|-|-|
| apple | 10 |
| banana | 12 |
`);
  expect(result).toMatchSnapshot();
});

test('images', () => {
  const transformer = new Transformer();
  const result = transformer.transform(`\
![](image1.png)

![](image2.png)
`);
  expect(result).toMatchSnapshot();
});

test('checkboxes', () => {
  const transformer = new Transformer();
  const result = transformer.transform(`\
# Housework

## Main

- [x] Dishes
- [ ] Cleaning the bathroom
- [x] Change the light bulbs
- [ ] something else

## [x] should it works on titles?

## [x] idk if it should!

### [ ] test

### [x] test

- [x] test
- [x] test


## [x] only works on list items is better
\`\`\`
[ ] this is not a checkbox either
\`\`\`
`);
  expect(result).toMatchSnapshot();
});

test('magic comments', () => {
  const transformer = new Transformer();
  const result = transformer.transform(`\
## heading 1 <!-- markmap: fold -->

- 1 <!-- markmap: foldAll -->
  - 1.1
  - 1.2
- 2
  - 2.1
  - 2.2
`);
  expect(result).toMatchSnapshot();
});

test('links - target=_blank', () => {
  const transformer = new Transformer([
    ...builtInPlugins,
    {
      name: 'target-blank',
      transform(transformHooks) {
        transformHooks.parser.tap((md) => {
          md.renderer.renderAttrs = wrapFunction(
            md.renderer.renderAttrs,
            (renderAttrs, token) => {
              let attrs = renderAttrs(token);
              if (token.type === 'link_open') {
                attrs += ' target="_blank"';
              }
              return attrs;
            },
          );
        });
        return {};
      },
    },
  ]);
  const result = transformer.transform(`\
## heading 1

- [Google](https://www.google.com)
`);
  expect(result).toMatchSnapshot();
});
