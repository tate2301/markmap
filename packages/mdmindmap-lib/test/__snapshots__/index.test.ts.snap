// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`checkboxes 1`] = `
{
  "content": "# Housework

## Main

- [x] Dishes
- [ ] Cleaning the bathroom
- [x] Change the light bulbs
- [ ] something else

## [x] should it works on titles?

## [x] idk if it should!

### [ ] test

### [x] test

- [x] test
- [x] test


## [x] only works on list items is better
\`\`\`
[ ] this is not a checkbox either
\`\`\`
",
  "features": {
    "hljs": true,
  },
  "parserOptions": undefined,
  "root": {
    "children": [
      {
        "children": [
          {
            "children": [],
            "content": "<svg width="16" height="16" viewBox="0 -3 24 24"><path d="M19 3H5a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2V5a2 2 0 0 0-2-2m-9 14-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8z"></path></svg> Dishes",
            "payload": {
              "lines": "4,5",
              "tag": "li",
            },
          },
          {
            "children": [],
            "content": "<svg width="16" height="16" viewBox="0 -3 24 24"><path fill-rule="evenodd" d="M6 5a1 1 0 0 0-1 1v12a1 1 0 0 0 1 1h12a1 1 0 0 0 1-1V6a1 1 0 0 0-1-1zM3 6a3 3 0 0 1 3-3h12a3 3 0 0 1 3 3v12a3 3 0 0 1-3 3H6a3 3 0 0 1-3-3v-5z" clip-rule="evenodd"></path></svg> Cleaning the bathroom",
            "payload": {
              "lines": "5,6",
              "tag": "li",
            },
          },
          {
            "children": [],
            "content": "<svg width="16" height="16" viewBox="0 -3 24 24"><path d="M19 3H5a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2V5a2 2 0 0 0-2-2m-9 14-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8z"></path></svg> Change the light bulbs",
            "payload": {
              "lines": "6,7",
              "tag": "li",
            },
          },
          {
            "children": [],
            "content": "<svg width="16" height="16" viewBox="0 -3 24 24"><path fill-rule="evenodd" d="M6 5a1 1 0 0 0-1 1v12a1 1 0 0 0 1 1h12a1 1 0 0 0 1-1V6a1 1 0 0 0-1-1zM3 6a3 3 0 0 1 3-3h12a3 3 0 0 1 3 3v12a3 3 0 0 1-3 3H6a3 3 0 0 1-3-3v-5z" clip-rule="evenodd"></path></svg> something else",
            "payload": {
              "lines": "7,9",
              "tag": "li",
            },
          },
        ],
        "content": "Main",
        "payload": {
          "lines": "2,3",
          "tag": "h2",
        },
      },
      {
        "children": [],
        "content": "<svg width="16" height="16" viewBox="0 -3 24 24"><path d="M19 3H5a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2V5a2 2 0 0 0-2-2m-9 14-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8z"></path></svg> should it works on titles?",
        "payload": {
          "lines": "9,10",
          "tag": "h2",
        },
      },
      {
        "children": [
          {
            "children": [],
            "content": "<svg width="16" height="16" viewBox="0 -3 24 24"><path fill-rule="evenodd" d="M6 5a1 1 0 0 0-1 1v12a1 1 0 0 0 1 1h12a1 1 0 0 0 1-1V6a1 1 0 0 0-1-1zM3 6a3 3 0 0 1 3-3h12a3 3 0 0 1 3 3v12a3 3 0 0 1-3 3H6a3 3 0 0 1-3-3v-5z" clip-rule="evenodd"></path></svg> test",
            "payload": {
              "lines": "13,14",
              "tag": "h3",
            },
          },
          {
            "children": [
              {
                "children": [],
                "content": "<svg width="16" height="16" viewBox="0 -3 24 24"><path d="M19 3H5a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2V5a2 2 0 0 0-2-2m-9 14-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8z"></path></svg> test",
                "payload": {
                  "lines": "17,18",
                  "tag": "li",
                },
              },
              {
                "children": [],
                "content": "<svg width="16" height="16" viewBox="0 -3 24 24"><path d="M19 3H5a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2V5a2 2 0 0 0-2-2m-9 14-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8z"></path></svg> test",
                "payload": {
                  "lines": "18,21",
                  "tag": "li",
                },
              },
            ],
            "content": "<svg width="16" height="16" viewBox="0 -3 24 24"><path d="M19 3H5a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2V5a2 2 0 0 0-2-2m-9 14-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8z"></path></svg> test",
            "payload": {
              "lines": "15,16",
              "tag": "h3",
            },
          },
        ],
        "content": "<svg width="16" height="16" viewBox="0 -3 24 24"><path d="M19 3H5a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2V5a2 2 0 0 0-2-2m-9 14-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8z"></path></svg> idk if it should!",
        "payload": {
          "lines": "11,12",
          "tag": "h2",
        },
      },
      {
        "children": [
          {
            "children": [],
            "content": "<pre data-lines="22,25"><code data-lines="22,25"><span class="hljs-string">[ ]</span> <span class="hljs-keyword">this</span> <span class="hljs-keyword">is</span> <span class="hljs-keyword">not</span> a checkbox either
</code></pre>",
            "payload": {
              "lines": "22,25",
              "tag": "pre",
            },
          },
        ],
        "content": "<svg width="16" height="16" viewBox="0 -3 24 24"><path d="M19 3H5a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2V5a2 2 0 0 0-2-2m-9 14-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8z"></path></svg> only works on list items is better",
        "payload": {
          "lines": "21,22",
          "tag": "h2",
        },
      },
    ],
    "content": "Housework",
    "payload": {
      "lines": "0,1",
      "tag": "h1",
    },
  },
}
`;

exports[`content with frontmatter 1`] = `
{
  "content": "---
markmap:
  color: blue
---

- l1
  - l1.1
  - l1.2
    - l1.2.1
",
  "features": {},
  "frontmatter": {
    "markmap": {
      "color": [
        "blue",
      ],
    },
  },
  "frontmatterInfo": {
    "lines": 4,
    "offset": 31,
  },
  "parserOptions": {},
  "root": {
    "children": [
      {
        "children": [],
        "content": "l1.1",
        "payload": {
          "lines": "6,7",
          "tag": "li",
        },
      },
      {
        "children": [
          {
            "children": [],
            "content": "l1.2.1",
            "payload": {
              "lines": "8,9",
              "tag": "li",
            },
          },
        ],
        "content": "l1.2",
        "payload": {
          "lines": "7,9",
          "tag": "li",
        },
      },
    ],
    "content": "l1",
    "payload": {
      "lines": "5,9",
      "tag": "li",
    },
  },
}
`;

exports[`content with line endings of CRLF 1`] = `
{
  "content": "---
markmap:
  color: blue
---

- l1
  - l1.1
  - l1.2
    - l1.2.1
",
  "features": {},
  "frontmatter": {
    "markmap": {
      "color": [
        "blue",
      ],
    },
  },
  "frontmatterInfo": {
    "lines": 4,
    "offset": 35,
  },
  "parserOptions": {},
  "root": {
    "children": [
      {
        "children": [],
        "content": "l1.1",
        "payload": {
          "lines": "6,7",
          "tag": "li",
        },
      },
      {
        "children": [
          {
            "children": [],
            "content": "l1.2.1",
            "payload": {
              "lines": "8,9",
              "tag": "li",
            },
          },
        ],
        "content": "l1.2",
        "payload": {
          "lines": "7,9",
          "tag": "li",
        },
      },
    ],
    "content": "l1",
    "payload": {
      "lines": "5,9",
      "tag": "li",
    },
  },
}
`;

exports[`content with only katex enabled 1`] = `
{
  "content": "---
markmap:
  color: blue
---

- $x = {-b \\pm \\sqrt{b^2-4ac} \\over 2a}$
",
  "features": {
    "katex": true,
  },
  "frontmatter": {
    "markmap": {
      "color": [
        "blue",
      ],
    },
  },
  "frontmatterInfo": {
    "lines": 4,
    "offset": 31,
  },
  "parserOptions": {},
  "root": {
    "children": [],
    "content": "<span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>x</mi><mo>=</mo><mfrac><mrow><mo>−</mo><mi>b</mi><mo>±</mo><msqrt><mrow><msup><mi>b</mi><mn>2</mn></msup><mo>−</mo><mn>4</mn><mi>a</mi><mi>c</mi></mrow></msqrt></mrow><mrow><mn>2</mn><mi>a</mi></mrow></mfrac></mrow><annotation encoding="application/x-tex">x = {-b \\pm \\sqrt{b^2-4ac} \\over 2a}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.4306em;"></span><span class="mord mathnormal">x</span><span class="mspace" style="margin-right:0.2778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2778em;"></span></span><span class="base"><span class="strut" style="height:1.3845em;vertical-align:-0.345em;"></span><span class="mord"><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.0395em;"><span style="top:-2.655em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">2</span><span class="mord mathnormal mtight">a</span></span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.394em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">−</span><span class="mord mathnormal mtight">b</span><span class="mbin mtight">±</span><span class="mord sqrt mtight"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.9221em;"><span class="svg-align" style="top:-3em;"><span class="pstrut" style="height:3em;"></span><span class="mord mtight" style="padding-left:0.833em;"><span class="mord mtight"><span class="mord mathnormal mtight">b</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.7463em;"><span style="top:-2.786em;margin-right:0.0714em;"><span class="pstrut" style="height:2.5em;"></span><span class="sizing reset-size3 size1 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span><span class="mbin mtight">−</span><span class="mord mtight">4</span><span class="mord mathnormal mtight">a</span><span class="mord mathnormal mtight">c</span></span></span><span style="top:-2.8821em;"><span class="pstrut" style="height:3em;"></span><span class="hide-tail mtight" style="min-width:0.853em;height:1.08em;"><svg xmlns="http://www.w3.org/2000/svg" width="400em" height="1.08em" viewBox="0 0 400000 1080" preserveAspectRatio="xMinYMin slice"><path d="M95,702
c-2.7,0,-7.17,-2.7,-13.5,-8c-5.8,-5.3,-9.5,-10,-9.5,-14
c0,-2,0.3,-3.3,1,-4c1.3,-2.7,23.83,-20.7,67.5,-54
c44.2,-33.3,65.8,-50.3,66.5,-51c1.3,-1.3,3,-2,5,-2c4.7,0,8.7,3.3,12,10
s173,378,173,378c0.7,0,35.3,-71,104,-213c68.7,-142,137.5,-285,206.5,-429
c69,-144,104.5,-217.7,106.5,-221
l0 -0
c5.3,-9.3,12,-14,20,-14
H400000v40H845.2724
s-225.272,467,-225.272,467s-235,486,-235,486c-2.7,4.7,-9,7,-19,7
c-6,0,-10,-1,-12,-3s-194,-422,-194,-422s-65,47,-65,47z
M834 80h400000v40h-400000z"></path></svg></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.1179em;"><span></span></span></span></span></span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.345em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span></span></span></span>",
    "payload": {
      "lines": "5,6",
      "tag": "li",
    },
  },
}
`;

exports[`content with only katex enabled 2`] = `
{
  "scripts": [
    {
      "data": {
        "fn": [Function],
        "getParams": [Function],
      },
      "type": "iife",
    },
    {
      "data": {
        "defer": true,
        "src": "https://cdn.jsdelivr.net/npm/webfontloader@1.6.28/webfontloader.js",
      },
      "type": "script",
    },
  ],
  "styles": [
    {
      "data": {
        "href": "https://cdn.jsdelivr.net/npm/katex@0.16.18/dist/katex.min.css",
      },
      "type": "stylesheet",
    },
  ],
}
`;

exports[`content without frontmatter 1`] = `
{
  "content": "- l1
  - l1.1
  - l1.2
    - l1.2.1
",
  "features": {},
  "parserOptions": undefined,
  "root": {
    "children": [
      {
        "children": [],
        "content": "l1.1",
        "payload": {
          "lines": "1,2",
          "tag": "li",
        },
      },
      {
        "children": [
          {
            "children": [],
            "content": "l1.2.1",
            "payload": {
              "lines": "3,4",
              "tag": "li",
            },
          },
        ],
        "content": "l1.2",
        "payload": {
          "lines": "2,4",
          "tag": "li",
        },
      },
    ],
    "content": "l1",
    "payload": {
      "lines": "0,4",
      "tag": "li",
    },
  },
}
`;

exports[`custom url provider 1`] = `
{
  "scripts": [
    {
      "data": {
        "fn": [Function],
        "getParams": [Function],
      },
      "type": "iife",
    },
    {
      "data": {
        "defer": true,
        "src": "/local/webfontloader@1.6.28/webfontloader.js",
      },
      "type": "script",
    },
  ],
  "styles": [
    {
      "data": {
        "href": "/local/katex@0.16.18/dist/katex.min.css",
      },
      "type": "stylesheet",
    },
    {
      "data": {
        "href": "/local/@highlightjs/cdn-assets@11.11.1/styles/default.min.css",
      },
      "type": "stylesheet",
    },
  ],
}
`;

exports[`images 1`] = `
{
  "content": "![](image1.png)

![](image2.png)
",
  "features": {},
  "parserOptions": undefined,
  "root": {
    "children": [
      {
        "children": [],
        "content": "<img src="image1.png" alt="">",
        "payload": {
          "lines": "0,1",
          "tag": "img",
        },
      },
      {
        "children": [],
        "content": "<img src="image2.png" alt="">",
        "payload": {
          "lines": "2,3",
          "tag": "img",
        },
      },
    ],
    "content": "",
  },
}
`;

exports[`links - target=_blank 1`] = `
{
  "content": "## heading 1

- [Google](https://www.google.com)
",
  "features": {},
  "parserOptions": undefined,
  "root": {
    "children": [
      {
        "children": [],
        "content": "<a href="https://www.google.com" target="_blank">Google</a>",
        "payload": {
          "lines": "2,3",
          "tag": "li",
        },
      },
    ],
    "content": "heading 1",
    "payload": {
      "lines": "0,1",
      "tag": "h2",
    },
  },
}
`;

exports[`magic comments 1`] = `
{
  "content": "## heading 1 <!-- markmap: fold -->

- 1 <!-- markmap: foldAll -->
  - 1.1
  - 1.2
- 2
  - 2.1
  - 2.2
",
  "features": {},
  "parserOptions": undefined,
  "root": {
    "children": [
      {
        "children": [
          {
            "children": [],
            "content": "1.1",
            "payload": {
              "lines": "3,4",
              "tag": "li",
            },
          },
          {
            "children": [],
            "content": "1.2",
            "payload": {
              "lines": "4,5",
              "tag": "li",
            },
          },
        ],
        "content": "1",
        "payload": {
          "fold": 2,
          "lines": "2,5",
          "tag": "li",
        },
      },
      {
        "children": [
          {
            "children": [],
            "content": "2.1",
            "payload": {
              "lines": "6,7",
              "tag": "li",
            },
          },
          {
            "children": [],
            "content": "2.2",
            "payload": {
              "lines": "7,8",
              "tag": "li",
            },
          },
        ],
        "content": "2",
        "payload": {
          "lines": "5,8",
          "tag": "li",
        },
      },
    ],
    "content": "heading 1",
    "payload": {
      "fold": 1,
      "lines": "0,1",
      "tag": "h2",
    },
  },
}
`;

exports[`plugins 1`] = `
{
  "scripts": [
    {
      "data": {
        "fn": [Function],
        "getParams": [Function],
      },
      "type": "iife",
    },
    {
      "data": {
        "defer": true,
        "src": "https://cdn.jsdelivr.net/npm/webfontloader@1.6.28/webfontloader.js",
      },
      "type": "script",
    },
  ],
  "styles": [
    {
      "data": {
        "href": "https://cdn.jsdelivr.net/npm/katex@0.16.18/dist/katex.min.css",
      },
      "type": "stylesheet",
    },
    {
      "data": {
        "href": "https://cdn.jsdelivr.net/npm/@highlightjs/cdn-assets@11.11.1/styles/default.min.css",
      },
      "type": "stylesheet",
    },
  ],
}
`;

exports[`tables 1`] = `
{
  "content": "| products | price |
|-|-|
| apple | 10 |
| banana | 12 |
",
  "features": {},
  "parserOptions": undefined,
  "root": {
    "children": [],
    "content": "<table data-lines="0,4">
<thead data-lines="0,1">
<tr data-lines="0,1">
<th>products</th>
<th>price</th>
</tr>
</thead>
<tbody data-lines="2,4">
<tr data-lines="2,3">
<td>apple</td>
<td>10</td>
</tr>
<tr data-lines="3,4">
<td>banana</td>
<td>12</td>
</tr>
</tbody>
</table>",
    "payload": {
      "lines": "0,4",
      "tag": "table",
    },
  },
}
`;
